control-plane/agents/src/bin/core/controller/scheduling/volume_policy/pool.rs

```diff
@@ -73,40 +73,88 @@ impl PoolBaseFilters {
-            Ok(spec) => match spec.labels {
-                None => false,
-                Some(label) => volume_pool_topology_labels
-                    .iter()
-                    .all(|(vol_key, vol_val)| {
-                        // See `InclusiveLabel` doc comment.
-                        // todo: add exclusion
-                        label
-                            .get(vol_key)
-                            .is_some_and(|pool_value| vol_val.is_empty() || pool_value == vol_val)
-                    }),
-            },
+            Ok(spec) => {
+                match spec.labels {
+                    None => false,
+                    Some(label) => {
+                        // The inclusion labes of volume must match the labels in pool
+                        // for the specified pool to qualify for volume provisioning.
+                        // i.e both the maps volume_pool_topology_inclusion_labels and label must be
+                        // equal
+                        let inclusion_match = label.iter().all(|(key, value)| {
+                            volume_pool_topology_inclusion_labels.get(key) == Some(value)
+                        }) && volume_pool_topology_inclusion_labels
+                            .iter()
+                            .all(|(key, value)| label.get(key) == Some(value));
+
+                        // The exclusion labels value in volume should not match the pool
+                        // labels value
+                        let common_keys: Vec<_> = volume_pool_topology_exclusion_labels
+                            .keys()
+                            .filter(|&key| label.contains_key(key))
+                            .collect();
+                        let exclusion_match = common_keys.iter().any(|&key| {
+                            volume_pool_topology_exclusion_labels.get(key) != label.get(key)
+                        });
+                        // removing the auto added key so that match can be done on user provided
+                        // values
+                        volume_pool_topology_inclusion_labels.remove("openebs.io/created-by");
+
+                        match (
+                            volume_pool_topology_inclusion_labels.is_empty(),
+                            volume_pool_topology_exclusion_labels.is_empty(),
+                        ) {
+                            (false, false) => {
+                                return inclusion_match && exclusion_match;
+                            }
+                            (false, true) => {
+                                return inclusion_match;
+                            }
+                            (true, false) => {
+                                return exclusion_match;
+                            }
+                            _ => {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
can we split the bits above this into two functions, one for inclusion and one for exclusion?

Reply:
removes all these logic and replace with does_pool_qualify_inclusion_labels and does_pool_qualify_exclusion_labels functions